tests/functional/test_snap_dcgm.py

```diff
@@ -0,0 +1,66 @@
+import json
+import subprocess
+import urllib.request
+from time import sleep
+
+from tenacity import retry, stop_after_delay, wait_fixed
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def test_dcgm_exporter():
+    """Test of the dcgm-exporter service and its endpoint."""
+    endpoint = "http://localhost:9400/metrics"
+
+    # Check the exporter endpoint, will raise an exception if the endpoint is not reachable
+    urllib.request.urlopen(endpoint)
+
+
+def test_dcgm_nv_hostengine():
+    """Check the dcgm-nv-hostengine service."""
+    nv_hostengine_service = "dcgm.nv-hostengine"
+
+    service = subprocess.run(
+        f"snap services {nv_hostengine_service}".split(),
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+
+    assert " active" in service.stdout.strip(), f"{nv_hostengine_service} service is not active"
+
+
+def test_dcgmi():
+    """Test of the dcgmi command."""
+    result = subprocess.run(
+        "dcgm.dcgmi discovery -l".split(), check=True, capture_output=True, text=True
+    )
+    assert "GPU ID" in result.stdout.strip(), "DCGMI is not working"
+
+
+def test_dcgm_bind_configs():
+    """Test snap port configuratin."""
+    services = ["dcgm.dcgm-exporter", "dcgm.nv-hostengine"]
+    configs = ["dcgm-exporter-address", "nv-hostengine-port"]
+    new_values = [":9466", "5666"]
+
+    result = subprocess.run(
+        "sudo snap get dcgm -d".split(), check=True, capture_output=True, text=True
+    )
+    dcgm_snap_config = json.loads(result.stdout.strip())
+    assert all(config in dcgm_snap_config for config in configs), "Missing snap configuration keys"
+
+    for config, new_value in zip(configs, new_values):
+        subprocess.run(
+            f"sudo snap set dcgm {config}={new_value}".split(), check=True
+        ), f"Failed to set {config}"
+
+    # restart the service to apply the new configuration
+    for service in services:
+        subprocess.run(f"sudo snap restart {service}".split(), check=True)
+
+    sleep(5)
```

Comment:
5 seconds is a significant delay, but you can't really rely on it always being enough, so this could lead to flaky tests. According to the docs `systemctl restart` does wait for the service to have restarted, did you observe it not being enough?